Add authentication confirmation endpoint

miniengineer opened the pull request.

zetter commented on config/routes.rb:
```diff
@@ -7,6 +7,12 @@
 
+  namespace :api do
+    namespace :v1 do
+      get '/me', to: 'profile#index'
```
You asked me about naming,

I think it depends on the concept your modelling - If you want to create an endpoint that returns information about the user (such as their user id of if they are signed in or not) then `me` sounds like a good name (even if your just using this data to check if they are signed in for now).

If you want to create and endpoint that just returns the authentication status then `/signInCheck` or `/authorizationCheck` might be clearer names. I think I've seen `/ping` used for similar purposes when the client regularly checks if the server is reachable.

miniengineer replied:
@zetter 
interesting, thank you 🙏 

I would probably leave the `/me` endpoint for the `#show` action.
it will return the user info + lunch_places, so it's better suited for the `/me` naming

And I will rename the `#index` to `/authenticationCheck` as you suggested to separate the logic 💯

zetter commented on app/javascript/src/components/Signup.js:
```diff
@@ -48,7 +48,7 @@ const Signup = () => {
 
     if (response.statusText === 'Created') {
-      setSignedIn(true);
+      setSignedIn('true');
```
I think I understand why you changed `true` to the string `'true'` here - it's so you represent the 3 states you need.

This is setting a trap for anyone using `isSignedIn` in the future - you might write `if (isSignedIn)` thinking it's a boolean but it's a string (and the string `'false'` is true!).

One option is to model it as states - change this to `setSignedState` with the possible values of `'signed_in'`, `'signed_out'`, or ``'unknown'`.

Another is to go back to `setSignedIn` setting booleans but change the values `UserContext` provides -

`<UserContext.Provider value={{ isSignedIn, setSignedIn }}>` could become something like:
`<UserContext.Provider value={{ isSignedIn, isSignedOut, isUnknown, setSignedIn }}>`

This code using `UserContext` will always be dealing with booleans.

miniengineer replied:
Thank you, this helps a lot.

Definitely, I was very close to falling into this trap
Will refactor to have all three states for `isSignedIn` as strings 👍

miniengineer commented:
@zetter 
thank you for the review 🙏 

I added a couple of changes

- renamed `/me` endpoint to `/authentication_check` 
- change `isSignedIn` to `signedState` (a bit lame, but the best I could come up with 😆 ) since to avoid any confusion with boolean
- `signedState` now has three states: `unknown`, `signed_in`, `signed_out`
- check the response status code instead of `statusText` - it's absent in case of sign out call for example
- added `email, setEmail` to `UserContext`
- added Sign out button component and included it into Profile page

Also have two questions 🤔 

1. I add `Accept: application/json` header for each request. From the [axios docs](https://axios-http.com/docs/config_defaults) I assumed that it's a common [default header](https://github.com/axios/axios/blob/master/lib/defaults.js#L121).
But it didn't work when I removed it from request options, so I made a wrapper function to follow DRY


2. I tried to use [webpack dev server](https://github.com/rails/webpacker/tree/5-x-stable#development) to instantly reload on React updates.
After running `./bin/webpack-dev-server` I got a question
`The command moved into a separate package: @webpack-cli/serve
Would you like to install serve? (That will run yarn add -D @webpack-cli/serve) `

After installing the `/serve` library, whenever I try to start webpack dev server I get the following error 👇 

```
✗ ./bin/webpack-dev-server
        return func(...args);
               ^
TypeError: Class constructor ServeCommand cannot be invoked without 'new'

```

It seems to be a common error and has to do with webpacker version.
Wanted to ask whether there is a better option other than degrading my webpacker? 🤔 
Thank you 🙏

zetter commented on app/javascript/src/components/App.js:
```diff
@@ -21,9 +21,10 @@ const App = () => {
+  if (signedStatus === 'unknown') return null;
 
-  return isSignedIn ? <Route {...props} /> : <Redirect to="/sign_in" />;
+  return signedStatus === 'signed_in' ? <Route {...props} /> : <Redirect to="/sign_in" />;
```
A small thing - `signedInStatus` might be clearer `signedStatus` since signing on it's own can refer to a different thing. Sorry if this came from a suggestion from me, I might have miss-typed it.

miniengineer replied:
I always have issues with straightforward variable naming 😆 
thank you!! will rename!!

zetter commented on config/routes.rb:
```diff
@@ -7,6 +7,12 @@
 
+  namespace :api do
+    namespace :v1 do
+      get '/authentication-check', to: 'profile#index'
```
Now you've decided this action is just about checking authentication, it could live in an `AuthenticationCheck`controller. Waiting until introducing other actions and controllers might help you decide the best place to put it.

zetter commented on app/javascript/src/utils/authentication/api.js:
```diff
@@ -1,10 +1,14 @@
 import { request } from 'axios';
 
+const baseRequest = (options) => {
+  options.headers = { Accept: 'application/json' };
+  return request(options);
+};
```
Axios does come with a suggested way of providing defaults (this code works where your using it now, but it overwrites all the headers so you can no longer provider your own if you need to).

https://axios-http.com/docs/instance

zetter commented:
Also have two questions 🤔

> I add Accept: application/json header for each request. From the axios docs I assumed that it's a common default header.
> But it didn't work when I removed it from request options, so I made a wrapper function to follow DRY

Great idea. I don't know why it didn't work - did you try to understand why?- is it that Axios isn't providing default headers to the request, or is it that the default headers don't work with Rails (since it lists other content types). 


> I tried to use webpack dev server to instantly reload on React updates.
> After running ./bin/webpack-dev-server I got a question
> The command moved into a separate package: @webpack-cli/serve Would you like to install serve? (That will run yarn add -D @webpack-cli/serve)

Yes, you might have to fix this by trying different versions of webpack (or webpack-cli). I might have time to look this weekend if your still having trouble. Changing webpack versions can be tricky to do and might have other compatibility problems with Rails - you might be able to find the version others use or a new install of Rails has.
